src/rgw/CMakeLists.txt

```diff
@@ -436,47 +344,326 @@ target_include_directories(radosgw
+  driver/rados/rgw_trim_mdlog.cc
+  driver/rados/rgw_user.cc
+  driver/rados/rgw_zone.cc
+  driver/rados/config/impl.cc
+  driver/rados/config/period.cc
+  driver/rados/config/period_config.cc
+  driver/rados/config/realm.cc
+  driver/rados/config/store.cc
+  driver/rados/config/zone.cc
+  driver/rados/config/zonegroup.cc)
 
-# this is unsatisfying and hopefully temporary; ARROW should not be
-# part of radosgw_admin
-if(WITH_RADOSGW_ARROW_FLIGHT)
-  list(APPEND radosgw_admin_srcs rgw_flight.cc)
-endif(WITH_RADOSGW_ARROW_FLIGHT)
+add_library(rados_sal OBJECT ${rados_sal_srcs})
+target_include_directories(rados_sal
+  PUBLIC "services"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/rgw"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/rgw/driver/rados"
+  PUBLIC "${LUA_INCLUDE_DIR}"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/dmclock/support/src")
 
-add_executable(radosgw-admin ${radosgw_admin_srcs})
-target_link_libraries(radosgw-admin ${rgw_libs} librados
+if(WITH_RADOSGW_KAFKA_ENDPOINT)
+  # used by rgw_kafka.cc
+  target_link_libraries(rados_sal
+    PRIVATE
+      RDKafka::RDKafka)
+endif()
+if(WITH_RADOSGW_AMQP_ENDPOINT)
+  # used by rgw_amqp.cc
+  target_link_libraries(rados_sal
+    PRIVATE
+      RabbitMQ::RabbitMQ
+      OpenSSL::SSL)
+endif()
+if(WITH_OPENLDAP)
+  target_link_libraries(rados_sal
+    PRIVATE
+      OpenLDAP::OpenLDAP)
+endif()
+if(WITH_RADOSGW_LUA_PACKAGES)
+  target_link_libraries(rados_sal
+    PRIVATE Boost::filesystem StdFilesystem::filesystem)
+endif()
+
+if(WITH_LTTNG)
+  # rgw/rgw_op.cc includes "tracing/rgw_op.h"
+  # rgw/rgw_rados.cc includes "tracing/rgw_rados.h"
+  add_dependencies(rados_sal rgw_op-tp rgw_rados-tp)
+endif()
+
+if(WITH_JAEGER)
+  add_dependencies(rados_sal jaeger_base)
+  target_link_libraries(rados_sal PUBLIC jaeger_base)
+endif()
+
+target_link_libraries(rados_sal
+  PUBLIC 
+    rgw_common
+  PRIVATE
+    global
+    cls_2pc_queue_client
+    cls_cmpomap_client
+    cls_lock_client
+    cls_log_client
+    cls_otp_client
+    cls_refcount_client
+    cls_rgw_client
+    cls_rgw_gc_client
+    cls_timeindex_client
+    cls_user_client
+    cls_version_client
+    librados)
+set_target_properties(rados_sal PROPERTIES COMPILE_FLAGS "-fPIC")
+
+
+set(rgw_sal_rados_srcs rgw_sal_rados_plugin.cc)
+add_library(rgw_sal_rados SHARED $<TARGET_OBJECTS:rados_sal> ${rgw_sal_rados_srcs})
+target_include_directories(rgw_sal_rados
+  PUBLIC "services"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/rgw"
+  PUBLIC "${CMAKE_SOURCE_DIR}/src/rgw/driver/rados"
+  PUBLIC "${LUA_INCLUDE_DIR}")
+
+if(WITH_LTTNG)
+  # rgw/rgw_op.cc includes "tracing/rgw_op.h"
+  # rgw/rgw_rados.cc includes "tracing/rgw_rados.h"
+  add_dependencies(rgw_sal_rados rgw_op-tp rgw_rados-tp)
+endif()
+
+if(WITH_JAEGER)
+  add_dependencies(rgw_sal_rados jaeger_base)
+  target_link_libraries(rgw_sal_rados PUBLIC jaeger_base)
+endif()
+
+target_link_libraries(rgw_sal_rados PUBLIC rados_sal)
+
+install(TARGETS rgw_sal_rados DESTINATION ${CMAKE_INSTALL_LIBDIR})
```

Comment:
now that we're loading these under `plugin_dir`, we should change the install paths to `${CEPH_INSTALL_PKGLIBDIR}`

other plugins use subdirectories under this `plugin_dir`, so maybe we should too?
```
src/crypto/CMakeLists.txt:set(crypto_plugin_dir ${CEPH_INSTALL_PKGLIBDIR}/crypto)
src/erasure-code/CMakeLists.txt:set(erasure_plugin_dir ${CEPH_INSTALL_PKGLIBDIR}/erasure-code)
src/librbd/CMakeLists.txt:set(librbd_plugins_dir ${CEPH_INSTALL_PKGLIBDIR}/librbd)
src/compressor/CMakeLists.txt:set(compressor_plugin_dir ${CEPH_INSTALL_PKGLIBDIR}/compressor)
src/extblkdev/CMakeLists.txt:set(extblkdev_plugin_dir ${CEPH_INSTALL_PKGLIBDIR}/extblkdev)
```

Reply:
will be in next push

Reply:
regarding the cmake magic for subdirs, i worked through this for the h3 frontend in https://github.com/ceph/ceph/pull/48178:
```cmake
set_target_properties(rgw_h3_quiche PROPERTIES VERSION 1.0.0 SOVERSION 1
  LIBRARY_OUTPUT_DIRECTORY "${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/h3")
```